R/namespace_linter.R

```diff
@@ -0,0 +1,127 @@
+#' @describeIn linters check for missing packages and symbols in namespace calls.
+#'   Note that using \code{check_exports=TRUE} or \code{check_nonexports=TRUE} will
+#'   load packages used in user code so it could potentially change the global state.
+#' @param check_exports Check if \code{symbol} is exported from \code{namespace} in
+#'   \code{namespace::symbol} calls.
+#' @param check_nonexports Check if \code{symbol} exists in \code{namespace} in
+#'   \code{namespace:::symbol} calls.
+#' @export
+namespace_linter <- function(check_exports = TRUE, check_nonexports = TRUE) {
+  function(source_file) {
+    if (!length(source_file$parsed_content)) return(list())
+
+    xml <- source_file$xml_parsed_content
+
+    ns_nodes <- xml2::xml_find_all(xml, "//expr/*[self::NS_GET or self::NS_GET_INT]")
+
+    get_node_text <- function(nodes) {
+      text <- xml2::xml_text(nodes)
+      expr <- parse(text = text, keep.source = FALSE)
+      vapply(expr, as.character, character(1L))
+    }
+
+    pkg_nodes <- xml2::xml_find_all(ns_nodes, "preceding-sibling::*[1]")
+    sym_nodes <- xml2::xml_find_all(ns_nodes, "following-sibling::*[1]")
+    pkgs <- get_node_text(pkg_nodes)
+    ops <- xml2::xml_text(ns_nodes)
+    syms <- get_node_text(sym_nodes)
+
+    installed_packages <- .packages(all.available = TRUE)
```

Comment:
I think I would prefer if this were a linter with options (function factory), and we can supply `installed_packages` externally (defaulting to what's used here).

would make this linter more useful in a broader set of use cases

Reply:
OK. The most general form could be user supplied installed packages. The use case might be customized library location, or a limited set of packages to check against instead of all currently installed packages. I think this makes sense. Let me create a PR to make changes for both `namespace_linter` and `missing_package_linter`.

Reply:
As for `namespace_linter`, if some packages in user supplied packages are not actually installed or the library path is different from the current session that runs the linter, then `getNamespace()` and other namespace functions won't work so that the linter will think the symbols after `::` and `:::` are not available and complain. How do you think it should behave?